Fix for #3805 - wrap macro that starts with nested body blocks

davidBar-On opened the pull request:
Fixes #3805

For wrapping macros with body block that starts with more than one `{`, i.e. starts with "nested blocks". The issue cause was that the nested block(s) `{` were not considered for calculating the remaining line width.

ytmimi commented on src/macros.rs:
```diff
@@ -1219,7 +1219,14 @@ impl MacroBranch {
         }
 
         // 5 = " => {"
-        let mut result = format_macro_args(context, self.args.clone(), shape.sub_width(5)?)?;
+        let old_body = context.snippet(self.body).trim();
+        let prefix_width = 5 + if context.config.format_macro_bodies() {
+            old_body.find(|ch| ch != '{').unwrap_or(old_body.len()) // " => {{..."
+        } else {
+            0
+        };
```
If we hoist the declaration for `has_block_body` up from line 1248 could this be simplified as follows:

```rust
        let old_body = context.snippet(self.body).trim();
        let has_block_body = old_body.starts_with('{');
        let prefix_width = if has_block_body {
            // 6 = " => {{"
            6
        } else {
            // 5 = " => {"
            5
        };
```

davidBar-On replied:
Changed, but kept the `if context.config.format_macro_bodies()`:
```rust
let prefix_width = if has_block_body && context.config.format_macro_bodies() {
    // 6 = " => {{"
    6
} else {
    // 5 = " => {"
    5
};
```

ytmimi replied:
Does the bug still pop up if `format_macro_bodies=false`?

I'm not sure we should conditionally set the prefix width based on the `format_macro_bodies` config argument since setting the prefix width to 5 when there are two opening braces is incorrect, and it's what lead to the bug in the first place.

davidBar-On replied:
Yes, the bug still pop if `format_macro_bodies=false`.  From you explanation to the other comment about the macro body, I now better understand the intention, so I **removed** the `if context.config.format_macro_bodies()` as you suggested.

ytmimi replied:
Great! Apologies if it wasn't totally clear what I was trying to accomplish with the refactored code snippet above.

ytmimi commented on tests/source/issue-3805.rs:
```diff
@@ -0,0 +1,43 @@
+// rustfmt-format_macro_matchers: true
+
+// From original issue example - Line length 101
+macro_rules! test {
+($aasdfghj:expr, $qwertyuiop:expr, $zxcvbnmasdfghjkl:expr, $aeiouaeiouaeio:expr, $add:expr) => {{
+return;
+}};
+}
```
Can we add a test case where we have nested blocks in the macro body:

```rust
macro_rules! test {
($aasdfghj:expr, $qwertyuiop:expr, $zxcvbnmasdfghjkl:expr, $aeiouaeiouaeio:expr, $add:expr) => {{{{
return;
}}}};
}
```

or even some whitespace between the two opening braces

```rust
macro_rules! test {
($aasdfghj:expr, $qwertyuiop:expr, $zxcvbnmasdfghjkl:expr, $aeiouaeiouaeio:expr, $add:expr) => {    {
return;
}};
}
```

davidBar-On replied:
Added the test cases.  Note that in the output of first test case below, the 3rd and 4th `{` are starting new indented block, which is different than keeping the 1st and 2nd `{` together.  However I assume this is acceptable.
```rust
macro_rules! test {
    (
        $aasdfghj:expr, $qwertyuiop:expr, $zxcvbnmasdfghjkl:expr, $aeiouaeiouaeio:expr, $add:expr
    ) => {{
        {
            {
                return;
            }
        }
    }};
}
```

ytmimi replied:
Yes, that's the intended behavior. the first set of braces denotes the arms of the macro. The second set of braces are part of the code that this macro will generates and will be the outer most block. All other braces should be indented as they define new nested blocks within the macro.

ytmimi commented:
Thanks for your help on this! I left two minor comments inline. Let me know your thoughts on both points.

davidBar-On commented:
Accepted the suggested changes, with minor change to the first comment.  Note also my note about the formatted output of new nest case with multi `{}`.

ytmimi commented on src/macros.rs:
```diff
@@ -1219,7 +1219,17 @@ impl MacroBranch {
         }
 
         // 5 = " => {"
-        let mut result = format_macro_args(context, self.args.clone(), shape.sub_width(5)?)?;
```
can we remove this leading comment since we comment what 5 and 6 represent down bellow

davidBar-On replied:
Removed

ytmimi reviewed (approved):
Thanks for the help fixing this issue, and for applying the feedback. I think we're good to go here!

calebcartwright commented:
I'm good with this change, but it will need to be version gated

@davidBar-On is that something you'd be willing to do when you have time?

davidBar-On commented:
> I'm good with this change, but it will need to be version gated

Added the version gating.
Also rebased and consolidated all changes into one commit.

Based on version other gating code, I used `if context.config.version() == Version::Two`, although I think that it may be better to use `if context.config.version() != Version::One` which is also good for `Version::Three`.

ytmimi commented:
This one has been marked as ready-to-merge for a while, and the fix has been version gated. Everything continues to pass after rebasing so I think we're good to go here.
